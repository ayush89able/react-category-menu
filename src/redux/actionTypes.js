export const ADD_CATEGORY = "ADD_CATEGORY";
export const ADD_SUBCATEGORY = "ADD_SUBCATEGORY";
export const DELETE_CATEGORY = "DELETE_CATEGORY";
export const EDIT_CATEGORY = "EDIT_CATEGORY";
