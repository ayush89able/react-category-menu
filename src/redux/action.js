import * as actionTypes from "./actionTypes";

export function addCategory(payload) {
  return { type: actionTypes.ADD_CATEGORY, payload };
}

export function addCategorySync(payload) {
  return dispatch => {
    setTimeout(() => {
      dispatch(addCategory(payload));
    }, 5000);
  };
}

export function addSubCategory(payload, index) {
  return { type: actionTypes.ADD_SUBCATEGORY, payload, index };
}

export function deleteCategory(index) {
  return { type: actionTypes.DELETE_CATEGORY, index };
}

export function editCategoryFunction(payload, index) {
  return { type: actionTypes.EDIT_CATEGORY, payload, index };
}
