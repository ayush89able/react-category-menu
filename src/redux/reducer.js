let initialState;
initialState = [
  {
    id: 1,
    value: "4 Wheelers",
    subCategory: [
      {
        id: 1,
        value: "Trucks",
        subCategory: []
      },
      {
        id: 1,
        value: "Cars",
        subCategory: []
      }
    ]
  },
  {
    id: 2,
    value: "2 wheeleres",
    subCategory: [
      {
        id: 1,
        value: "Bikes",
        subCategory: []
      }
    ]
  },
  {
    id: 3,
    value: "Electronics",
    subCategory: []
  }
];
console.log(localStorage.getItem("state"));
if (localStorage.getItem("state") != null) {
  initialState = JSON.parse(localStorage.getItem("state"));
}

import * as actionTypes from "./actionTypes";

function catagories(state = initialState, action) {
  console.log("action", action);
  switch (action.type) {
    case actionTypes.ADD_CATEGORY:
      return state.concat([action.payload]);
    case actionTypes.ADD_SUBCATEGORY:
      return [
        // part of the array before the specified index
        ...state.slice(0, action.index),
        // inserted item
        action.payload,
        // part of the array after the specified index
        ...state.slice(action.index + 1)
      ];
    case actionTypes.DELETE_CATEGORY:
      let stateCopy = [...state];
      stateCopy.splice(action.index, 1);
      return stateCopy;
    case actionTypes.EDIT_CATEGORY:
      console.log("aasda", action.payload);
      return [
        // part of the array before the specified index
        ...state.slice(0, action.index),
        // inserted item
        action.payload,
        // part of the array after the specified index
        ...state.slice(action.index + 1)
      ];
    default:
      return state;
  }
}

export default catagories;
