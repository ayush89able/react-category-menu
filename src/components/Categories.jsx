import React, { useState } from 'react'
import { useSelector, useDispatch } from "react-redux";
import { addCategorySync, addSubCategory, deleteCategory, editCategoryFunction } from '../redux/action'
import { FaPlus, FaMinus, FaPlusSquare, FaMinusSquare, FaTrash, FaRegEdit } from "react-icons/fa";

const Categories = () => {

  const [newCategory, setNewCategory] = useState('')
  const [newSubCategory, setNewSubCategory] = useState('')
  const [editCategory, setEditCategory] = useState('')
  const [expandAll, setExpandAll] = useState(false)
  const [expand, setExpand] = useState(false)
  const [indexNumber, setIndexNumber] = useState(null)
  const [editIndexNumber, setEditIndexNumber] = useState(null)

  const state = useSelector(state => state);
  // console.log('state',state)
  const dispatch = useDispatch()

  const addCategoryHandler = () => {
    console.log('Adding Category')
    console.log(newCategory)
    let payload = {
      id: state.length + 1,
      value: newCategory,
      subCategory: []
    }
    console.log(payload)
    dispatch(addCategorySync(payload))
    setNewCategory('')
  }

  const showAddSubCategory = (index, item) => {
    setIndexNumber(index)
    setExpand(!expand)
  }

  const showEditCategory = (index, item) => {
    setEditIndexNumber(index)
  }

  const deleteCategoryHandler = (index, item) => {
    console.log('Deleteing Category')
    console.log('index', index)
    console.log('item', item)
    dispatch(deleteCategory(index))
  }

  const addSubCategoryHandler = (index, item) => {
    console.log('index', index)
    console.log('item', item)
    let subCategoryLength = item.subCategory.length
    console.log(subCategoryLength)
    let payload = {
      id: subCategoryLength + 1,
      value: newSubCategory,
      subCategory: []
    }
    let subCategory = item.subCategory.concat(payload);
    console.log('subCategories', subCategory)
    let payloadTwo = {
      ...item,
      subCategory
    }
    console.log('payloadTwo', payloadTwo)
    dispatch(addSubCategory(payloadTwo, index))
    setIndexNumber(null)
    setNewSubCategory('')
  }

  const editCategoryHandler = (index, item) => {
    console.log('editCategory', editCategory)
    console.log('index', index)
    console.log('item', item)
    let payload = {
      ...item,
      value: editCategory
    }
    console.log('payload edit Category', payload)
    dispatch(editCategoryFunction(payload, index))
    setEditIndexNumber(null)
    setEditCategory('')
  }

  const saveHandler = () => {
    localStorage.setItem('state', JSON.stringify(state))
  }


  return (
    <div className="container">
      <h1>{`${state.length} Categories Present`}</h1>
      <div className="row" style={{ marginBottom: '1rem' }}>
        <div className="col">
        </div>
        <div className="col">
          <button onClick={() => setExpandAll(!expandAll)} data-testid='expandAllButton' className="btn btn-primary">{expandAll ? <FaMinusSquare /> : <FaPlusSquare />}</button>
        </div>
        <div className="col">
          <button onClick={saveHandler} data-testid='saveToLocalButton' className="btn btn-primary">Save to local</button>
        </div>
        <div className="col">
        </div>
      </div>
      {state.map((item, index) => {
        return (
          <div className="row" key={index} style={{ marginBottom: '1em' }}>
            <div className="col">
              {item.value}
              {item.subCategory.length > 0 && expandAll && (
                item.subCategory.map((subItem, index) => {
                  return (
                    <div className="row" key={index} style={{ marginLeft: '1em' }}>
                      <div className="col">
                        <div>{subItem.value}</div>
                      </div>
                    </div>
                  )
                })
              )}
              {item.subCategory.length > 0 && indexNumber === index && expand && (
                item.subCategory.map((subItem, index) => {
                  return (
                    <div className="row" key={index} style={{ marginLeft: '1em' }}>
                      <div className="col">
                        <div>{subItem.value}</div>
                      </div>
                    </div>
                  )
                })
              )}
            </div>
            <div className="col">
              <button id={index} data-testid={`addSubCategoryOpenButton-${index}`} onClick={() => { showAddSubCategory(index, item) }} className="btn btn-primary">{expand && indexNumber === index ? <FaMinus /> : <FaPlus />}</button>
              {indexNumber === index && expand && (
                <div className="row" style={{ margin: '1em 0' }}>
                  <input type='text' value={newSubCategory} data-testid={`addSubCategoryInput-${index}`} placeholder='New Sub Category' onChange={(e) => setNewSubCategory(e.target.value)} />
                  <button id={index} data-testid={`addSubCategoryButton-${index}`} onClick={() => { addSubCategoryHandler(index, item) }} className="btn btn-primary" style={{ margin: '0.5em 0' }}>Add Sub Category</button>
                </div>
              )}
            </div>
            <div className="col">
              <button id={index} data-testid={`deleteButton-${index}`} onClick={() => { deleteCategoryHandler(index, item) }} className="btn btn-primary"><FaTrash /></button>
            </div>
            <div className="col">
              <button id={index} data-testid={`editCategoryOpenButton-${index}`} onClick={() => { showEditCategory(index, item) }} className="btn btn-primary"><FaRegEdit /></button>
              {editIndexNumber === index && (
                <div className="row" style={{ margin: '1em 0' }}>
                  <input placeholder='Edited Category Name' data-testid={`editCategoryInput-${index}`} type='text' value={editCategory} onChange={(e) => setEditCategory(e.target.value)} />
                  <button id={index} data-testid={`editCategoryButton-${index}`} onClick={() => { editCategoryHandler(index, item) }} className="btn btn-primary" style={{ margin: '0.5em 0' }}>Edit Category</button>
                </div>
              )}
            </div>
          </div>
        )
      })}
      <div className="row">
        <div className="col">
          <input type='text' placeholder='New Category' value={newCategory} onChange={(e) => setNewCategory(e.target.value)} />
        </div>
        <div className="col">
          <button data-testid='addCategoryButton' onClick={addCategoryHandler} className="btn btn-primary">Add Category</button>
        </div>
      </div>
    </div>
  )
}
export default Categories