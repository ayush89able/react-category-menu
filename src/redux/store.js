import { createStore } from "redux";
import catagories from "./reducer";

const store = createStore(catagories);

export default store;
