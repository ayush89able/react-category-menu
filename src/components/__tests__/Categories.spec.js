import React from 'react'
// import renderer from 'react-test-renderer';
import { Provider } from 'react-redux'
import Categories from '../Categories'
import { render, screen, fireEvent  } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { deleteCategory, addSubCategory, editCategoryFunction } from '../../redux/action'
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);

describe('Categories Component', () => {

    let store;
    let component;
    beforeEach(() => {
        let initialState = [
            {
                id: 1,
                value: "4 Wheelers",
                subCategory: [
                    {
                        id: 1,
                        value: "Trucks",
                        subCategory: []
                    },
                    {
                        id: 1,
                        value: "Cars",
                        subCategory: []
                    }
                ]
            },
            {
                id: 2,
                value: "2 wheeleres",
                subCategory: [
                    {
                        id: 1,
                        value: "Bikes",
                        subCategory: []
                    }
                ]
            },
            {
                id: 3,
                value: "Electronics",
                subCategory: []
            }
        ];

        store = mockStore(initialState);
        store.dispatch = jest.fn();

        // component = renderer.create(
        //     <Provider store={store}>
        //         <Categories />
        //     </Provider>
        // );
        const { container } = render(
            <Provider store={store}>
                <Categories />
            </Provider>
        );
        component = container
    });

    it('should render with given state from Redux store', () => {
        // expect(component.toJSON()).toMatchSnapshot();
        expect(component.firstChild).toMatchSnapshot();
    });

    it('should have three categories initially', () => {
        let cn = screen.getByTestId('categoriesNumber')
        expect(cn.innerHTML).toBe("3 Categories Present");
    });

    it('should dispatch an action on add sub category button click', () => {
        userEvent.click(screen.getByTestId('addSubCategoryOpenButton-0'))
        userEvent.click(screen.getByTestId('addSubCategoryButton-0'))
        expect(store.dispatch).toHaveBeenCalledTimes(1);
    });

    it('should dispatch an action on edit category button click', () => {
        userEvent.click(screen.getByTestId('editCategoryOpenButton-0'))
        userEvent.click(screen.getByTestId('editCategoryButton-0'))
        expect(store.dispatch).toHaveBeenCalledTimes(1);
    });

    it('should dispatch an action on delete button click', () => {
        userEvent.click(screen.getByTestId('deleteButton-0'))
        expect(store.dispatch).toHaveBeenCalledTimes(1);
    }); 

    it('should dispatch an action on add category button click', () => {
        userEvent.click(screen.getByTestId('addCategoryButton'))
        expect(store.dispatch).toHaveBeenCalledTimes(1);       
    });

    it('should dispatch an action on saveToLocalButton button click', () => {
        userEvent.click(screen.getByTestId('saveToLocalButton'))
        expect(component.firstChild).toMatchSnapshot();
    });

    it('should dispatch an action on expandAllButton button click', () => {
        userEvent.click(screen.getByTestId('expandAllButton'))
        expect(component.firstChild).toMatchSnapshot();
    });

    // it('should decrease number of categories on delete event', () => {
    //     userEvent.click(screen.getByTestId('deleteButton-0'))
    //     let cn = screen.getByTestId('categoriesNumber')
    //     expect(cn.innerHTML).toBe("2 Categories Present")       
    // });

    // it('should increase number of categories on add category event after 5000ms', () => {
    //     let input = screen.getByTestId('addCategoryInput')
    //     fireEvent.change(input, { target: { value: 'Acsss' } })
    //     userEvent.click(screen.getByTestId('addCategoryButton'))
    //     let cn = screen.getByTestId('categoriesNumber')
    //     expect(cn).toBe("4 Categories Present");
    // });

})

describe('Action Creators',()=>{
    test('delete action creator action',()=>{
       const expectedAction = {
          type:'DELETE_CATEGORY',
          index: 0 
       }
       expect(deleteCategory(0)).toEqual(expectedAction);
    })

    test('addSubCategory action creator action',()=>{
        const expectedAction = {
           type:'ADD_SUBCATEGORY',
           index: 0 ,
           payload: {}
        }
        expect(addSubCategory({},0)).toEqual(expectedAction);
     })

     test('editCategoryFunction action creator action',()=>{
        const expectedAction = {
           type:'EDIT_CATEGORY',
           index: 0 ,
           payload: {}
        }
        expect(editCategoryFunction({},0)).toEqual(expectedAction);
     })
 })