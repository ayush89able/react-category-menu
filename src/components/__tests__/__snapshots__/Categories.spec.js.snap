// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Component should render with given state from Redux store 1`] = `
<div
  class="container"
>
  <h1>
    3 Categories Present
  </h1>
  <div
    class="row"
    style="margin-bottom: 1rem;"
  >
    <div
      class="col"
    />
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
      >
        Save to local
      </button>
    </div>
    <div
      class="col"
    />
  </div>
  <div
    class="row"
    style="margin-bottom: 1em;"
  >
    <div
      class="col"
    >
      4 Wheelers
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="addSubCategoryOpenButton-0"
        id="0"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="deleteButton-0"
        id="0"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="editCategoryOpenButton-0"
        id="0"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M402.3 344.9l32-32c5-5 13.7-1.5 13.7 5.7V464c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h273.5c7.1 0 10.7 8.6 5.7 13.7l-32 32c-1.5 1.5-3.5 2.3-5.7 2.3H48v352h352V350.5c0-2.1.8-4.1 2.3-5.6zm156.6-201.8L296.3 405.7l-90.4 10c-26.2 2.9-48.5-19.2-45.6-45.6l10-90.4L432.9 17.1c22.9-22.9 59.9-22.9 82.7 0l43.2 43.2c22.9 22.9 22.9 60 .1 82.8zM460.1 174L402 115.9 216.2 301.8l-7.3 65.3 65.3-7.3L460.1 174zm64.8-79.7l-43.2-43.2c-4.1-4.1-10.8-4.1-14.8 0L436 82l58.1 58.1 30.9-30.9c4-4.2 4-10.8-.1-14.9z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="row"
    style="margin-bottom: 1em;"
  >
    <div
      class="col"
    >
      2 wheeleres
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="addSubCategoryOpenButton-1"
        id="1"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="deleteButton-1"
        id="1"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="editCategoryOpenButton-1"
        id="1"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M402.3 344.9l32-32c5-5 13.7-1.5 13.7 5.7V464c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h273.5c7.1 0 10.7 8.6 5.7 13.7l-32 32c-1.5 1.5-3.5 2.3-5.7 2.3H48v352h352V350.5c0-2.1.8-4.1 2.3-5.6zm156.6-201.8L296.3 405.7l-90.4 10c-26.2 2.9-48.5-19.2-45.6-45.6l10-90.4L432.9 17.1c22.9-22.9 59.9-22.9 82.7 0l43.2 43.2c22.9 22.9 22.9 60 .1 82.8zM460.1 174L402 115.9 216.2 301.8l-7.3 65.3 65.3-7.3L460.1 174zm64.8-79.7l-43.2-43.2c-4.1-4.1-10.8-4.1-14.8 0L436 82l58.1 58.1 30.9-30.9c4-4.2 4-10.8-.1-14.9z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="row"
    style="margin-bottom: 1em;"
  >
    <div
      class="col"
    >
      Electronics
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="addSubCategoryOpenButton-2"
        id="2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="deleteButton-2"
        id="2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
          />
        </svg>
      </button>
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="editCategoryOpenButton-2"
        id="2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M402.3 344.9l32-32c5-5 13.7-1.5 13.7 5.7V464c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V112c0-26.5 21.5-48 48-48h273.5c7.1 0 10.7 8.6 5.7 13.7l-32 32c-1.5 1.5-3.5 2.3-5.7 2.3H48v352h352V350.5c0-2.1.8-4.1 2.3-5.6zm156.6-201.8L296.3 405.7l-90.4 10c-26.2 2.9-48.5-19.2-45.6-45.6l10-90.4L432.9 17.1c22.9-22.9 59.9-22.9 82.7 0l43.2 43.2c22.9 22.9 22.9 60 .1 82.8zM460.1 174L402 115.9 216.2 301.8l-7.3 65.3 65.3-7.3L460.1 174zm64.8-79.7l-43.2-43.2c-4.1-4.1-10.8-4.1-14.8 0L436 82l58.1 58.1 30.9-30.9c4-4.2 4-10.8-.1-14.9z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <input
        placeholder="New Category"
        type="text"
        value=""
      />
    </div>
    <div
      class="col"
    >
      <button
        class="btn btn-primary"
        data-testid="addCategoryButton"
      >
        Add Category
      </button>
    </div>
  </div>
</div>
`;
